### ISSUE TITLE
`set_dependency()` does not successfully check modes

### ISSUE THREAD

**Issue by juliasilge** [tidymodels/parsnip#630]
In the changes made in #629, there is code that checks the mode (like `!any(mode == all_modes)`) and is meant to error but it looks like it is not working correctly:

``` r
library(parsnip)

set_new_model("sponge")
set_model_mode("sponge", "regression")

## maybe this should error?
set_model_engine("sponge", mode = "classification", eng = "gum")       

# definitely should error based on new PR, right?
set_dependency("sponge",  "gum", "trident", mode = "classification")    
```

<sup>Created on 2022-01-03 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

None of that errors as of my writing this.

**Comment by topepo** [comment]
This is an issue with `check_mode_val()`. `set_model_engine()` should error since it must be upstream of `set_dependency()`: 

``` r
library(parsnip)

set_new_model("sponge")
set_model_mode("sponge", "regression")
set_dependency("sponge",  "gum", "trident", mode = "classification")   
#> Error: The engine 'gum' has not been registered for model 'sponge'.
```

<sup>Created on 2022-01-03 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by topepo** [comment]
It looks like `check_spec_mode_engine_val(()` was not the best fit to check these. I've added a new function that is specifically for what we are interested in.

**Comment by juliasilge** [comment]
Closed in #629
